Common/Ubuntu/ap_camera_non_gps/feature_match.py

```diff
@@ -0,0 +1,182 @@
+        src_pts = np.float32([ kp1[m.queryIdx].pt for m in good ]).reshape(-1,1,2)
+        dst_pts = np.float32([ kp2[m.trainIdx].pt for m in good ]).reshape(-1,1,2)
+        M, mask = cv.findHomography(src_pts, dst_pts, cv.RANSAC,5.0)
+        matchesMask = mask.ravel().tolist()
+        h,w = img1.shape
+        pts = np.float32([ [0,0],[0,h-1],[w-1,h-1],[w-1,0] ]).reshape(-1,1,2)
+        dst = cv.perspectiveTransform(pts,M)
+        img2 = cv.polylines(img2,[np.int32(dst)],True,255,3, cv.LINE_AA)
+
+
+
+        #Orientation
+
+        diff = np.abs(dst - pts)
+        orientation_diff = np.arctan2(np.abs(dst[1][0][1] - dst[0][0][1]), np.abs(dst[1][0][0] - dst[0][0][0]))
+        print("Orientation difference: {:.2f} degrees".format(90 - np.degrees(orientation_diff)))
+
+
+        #x, y offset
+
+        h1, w1 = img1.shape[:2]
+        h, w = img2.shape
+
+        sr_pts = np.float32([kp1[m.queryIdx].pt for m in good]).reshape(-1, 1, 2)
+        ds_pts = np.float32([kp2[m.trainIdx].pt for m in good]).reshape(-1, 1, 2)
+        Mr, maskr = cv.findHomography(ds_pts, sr_pts, cv.RANSAC, 5.0)
+
+        corners = np.array([[0, 0], [0, h-1], [w-1, h-1], [w-1, 0]], dtype=np.float32).reshape(-1, 1, 2)
+        transformed_corners = cv.perspectiveTransform(corners, Mr)
+
+        x, y = np.mean(transformed_corners, axis=0).astype(int)[0]
+
+        print("Offset x, y: ",x-w1/2, y-h1/2)
+
+        inCMX = (x-w1/2) * (10) / focal
+        inCMY = (y-h1/2) * (10) / focal
+
+
+        #Zoom percentage -- Not complete
+
+        width = abs(transformed_corners[2][0][0]) - abs(transformed_corners[0][0][0])
+        height = abs(transformed_corners[2][0][1]) - abs(transformed_corners[0][0][1])
+
+        im2 = width * height
+        im1 = img1.shape[0] * img1.shape[1]
+
+        #print(transformed_corners)
+
+        print(f"Zoom percent: {im2/im1*100} %")
+
+    else:
+        print( "Not enough matches are found - {}/{}".format(len(good), MIN_MATCH_COUNT) )
+        matchesMask = None
+
+
+    draw_params = dict(matchColor = (0,255,0),
+                       singlePointColor = (255,0,0),
+                       matchesMask = matchesMask,
+                       flags = 2)
+
+
+    binToBool= [True if n == 1 else False for n in matchesMask]
+    res = list(filter(lambda i: binToBool[i], range(len(binToBool))))
+    res_listQ = [src_pts[i].tolist() for i in res]
+    res_listT = [dst_pts[i].tolist() for i in res]
+
+
+    cv.namedWindow("img3", cv.WINDOW_NORMAL)
+    img3 = cv.drawMatches(img1,kp1,img2,kp2,good,None,**draw_params)
+
+    cv.imshow('img3', img3)
+
+    cv.waitKey(0)
+
+
+
+def offset_latlng(lat,lng,ofs_north,ofs_east):
+    dlat = ofs_north * LOCATION_SCALING_FACTOR_INV
+    dlng = (ofs_east * LOCATION_SCALING_FACTOR_INV) / longitude_scale(lat+dlat/2)
+    lat += dlat
+    lat = limit_lattitude(lat)
+    lng = wrap_longitude(dlng+lng)
+    print("New Lat & Lng: ",lat,lng)
+
+def MAX(one, two):
+    return one if one > two else two
```

Comment:
You can use this: https://www.w3schools.com/python/ref_func_max.asp

Reply:
Thank you sir! Updated